### ISSUE TITLE
GitLab API support.

### ISSUE THREAD

**Issue by cirosantilli** [prose/prose#713]
Would pull requests be considered to support the GitLab API to allow Prose to edit GitLab content?

GitLab also has a CRUD file API: https://github.com/gitlabhq/gitlabhq/blob/master/doc/api/repository_files.md

This was mentioned at https://github.com/prose/prose/issues/508#issuecomment-21742995, but we should have a separate issue for it.

The goal is to develop something like: https://github.com/booktree/booktree

This should be implemented as a portable API wrapper that uses a single interface for both GitHub and GitLab. If such wrapper existed and were used, support for other providers like Bitbucket would be simple as well https://github.com/prose/prose/issues/545

**Comment by fulldecent** [comment]
Is anyone inclined to work on this? Recommending to close this issue unless there is a potential chance of someone picking it up.

**Comment by purldator** [comment]
I too would like to see this feature.  Only alternative is to give in to Penflip, which isn't open source but [does use the same exact stack of GitLab and Prose](http://booktree.github.io/alternatives/).  Enabling Prose to make use of GitLab API functionality would enable anyone to create their own private Penflip instances in the cloud.  That would open the door to a lot of exciting prospects such as combining GitLab and Prose with Jekyll to make a fully-collaborative, versoning CMS or Wiki without relying on third party servers such as GitHub  [It is briefly touched upon here](http://eric.pdxhub.org/articles/jekyll-pages-prose).  Would be a shame if this issue was closed.

**Comment by cirosantilli** [comment]
@purldator Penflip is built on GitLab https://twitter.com/madebyloren/status/468917323588698112 and so is O'Reilly Atlas https://atlas.oreilly.com/ , another closed source one, so I don't think it makes much difference. I started considering such a project at: https://github.com/booktree/booktree

**Comment by purldator** [comment]
@cirosantilli Yes, that is where my first link leads from.  Booktree looks like a good alternative but for my uses I am not looking to publish in that measure.

The difference is Prose allows non-developers to interface with git (Via GitLab).  I've already tried to get a non-developer to use GitLab alone.  They refused as it was 'too complicated'.  Prose would mitigate that hard negative response.

**Comment by cirosantilli** [comment]
@purldator 

>  Prose would mitigate that hard negative response.

That is exactly my goal: paste a prose or something else in front of gitlab to make it easier on newbs :) Let me know if you have new results.

**Comment by mikemorris** [comment]
Looping in @aguynamedloren as he has [previously expressed interest in contributing back to Prose](https://github.com/prose/prose/issues/743#issuecomment-54594863). If Penflip already has a working GitLab backend, a pull request adding a [pluggable backend structure](https://github.com/prose/prose/issues/713#issuecomment-46985015) with GitHub and GitLab support could be a great contribution.

**Comment by timbur** [comment]
Any progress made on this?

**Comment by fbender** [comment]
If no one is working on this, I may have time* to write a PoC SCM abstraction tailored for existing Github and requested Gitlab integration. I'd be happy about any pointer or existing code snippet anyone may already have.

* Pending project management decision.

**Comment by bepetersn** [comment]
+1 I would also love to see Gitlab supported by prose, it would allow me to collaborate better with writers. Thanks much!

**Comment by pierreozoux** [comment]
Any update on this? Is it a big amount of work? Could we fund it on bountysource?

**Comment by dereklieu** [comment]
Update on this: [iilab](https://iilab.org/) sponsored some work relating to Prose and Gitlab. One of the items they sponsored was a Gitlab port. I have 90% of the work done in my fork [dereklieu/prose](https://github.com/dereklieu/prose) with the goal of merging it back into prose/prose sometime within the next few months. When it's ready I'll post an update in this ticket.

If you're curious on the progress and want to start using it now, it's still somewhat buggy and I can't offer you any support, but you're welcome to pull down my fork. You'll need to set up and configure your own instance of something like [prose/gatekeeper](https://github.com/prose/gatekeeper).

**Comment by Samsemilio** [comment]
@dereklieu 
Are there any updates on your integration of prose with gitlab?

**Comment by dereklieu** [comment]
No, I'm working on an elections-related project at the moment and won't have time to come up for air until after Nov 8.

**Comment by DarwinJS** [comment]
I would love to see this as well.

I noticed a couple people mentioned private repos on gitlab, but I didn't see it mentioned explicitly that you get them for free.

Being able to keep my site source private, but not pay a monthly to do so, is why I would love to see git lab support.

**Comment by attiks** [comment]
Private repos are free on gitlab.com

**Comment by DarwinJS** [comment]
My point precisely :)

**Comment by dereklieu** [comment]
I have a gitlab fork of prose deployed at [lieu.io/prose](http://lieu.io/prose/). The github repo is [here](https://github.com/dereklieu/prose). This is a *very very early beta* and some features like forking haven't been implemented, but the core editing workflow is in place. Feel free to try it and open issues specific to that fork at the repo I listed above.

**Comment by bnvk** [comment]
@dereklieu awesome, I'll give this a test. What features are missing?

**Comment by dereklieu** [comment]
Besides forking I honestly can't remember, it's been a while since I was in that code.

**Comment by ImaCrea** [comment]
Hi folks, would love to know more about this. Our project MailTape is heavily dependent of Prose but as we want to move to a private repo and are a non-profit organization Gitlab would be the solution to choose. Any alternative maybe?

**Comment by dereklieu** [comment]
@ImaCrea for now I invite you to try out the gitlab fork I referenced above, to see if it'll if it'll fit your needs.

**Comment by timwood** [comment]
@dereklieu Can I run your fork in my own environment and against my own Gitlab CE instance?

**Comment by dereklieu** [comment]
@timwood of course.

**Comment by caiofbpa** [comment]
Any news on stabilizing the GitLab support and merging it back to official prose?

Can I help in any way?

**Comment by b-long** [comment]
Ditto to @caiofbpa 's comment, I'd be happy to help test!

**Comment by StanLindsey** [comment]
I'd be interested in this also. What do we need to do to make this happen?

**Comment by ngustavo** [comment]
Any news on this? It has been more than 6 months.

**Comment by duckbrain** [comment]
@dereklieu's fork no longer works at all on the latest GitLab. It uses the v3 API, which has been removed in favor of v4. I'm going to see if I can get it rebased and working when I've got some time.

**Comment by bpmottathai** [comment]
+1 for this project, any update...?
